### PR TITLE
Use number of bins instead of quantization interval in mlos_bench tunables

### LINKED ISSUE #803: Implement quantization for ConfigSpace >= 1.0
ConfigSpace 1.0 hyperparameters do not have the quantization parameter `q=` anymore. In #802 we temporarily remove support for quantization in ConfigSpace. We need to add it back using the new `quantize()` family of functions of ConfigSpace.

### PR DESCRIPTION
**Pull request by motus**
Closes #803 

> Future PR will rename the config schema in order to reduce confusion on the change in semantics, but also keep this PR smaller.

### CODE REVIEW

**Comment by bpkroth** [comment]
> @bpkroth before you merge, can you please run tests locally for this PR? One unit test, `toy_optimization_loop_test.py::test_flaml_optimization_loop`, now produces results on GitHub that are different from what we've had before and what I am getting on my laptop. I have not had time to investigate the issue yet.

This is due to a bump in FLAML:
- #839
- #840

File: mlos_bench/mlos_bench/optimizers/convert_configspace.py [hunk 1]
```diff
@@ -149,6 +163,9 @@ def _tunable_to_configspace(
     else:
         raise TypeError(f"Invalid Parameter Type: {tunable.type}")
 
+    if tunable.quantization:
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/optimizers/convert_configspace.py:166]:
```suggestion
    if tunable.quantization:
        # Temporary workaround to dropped quantization support in ConfigSpace 1.0
        # See Also: https://github.com/automl/ConfigSpace/issues/390
```

File: mlos_bench/mlos_bench/optimizers/convert_configspace.py [hunk 2]
```diff
@@ -77,6 +99,7 @@ def _tunable_to_configspace(
     cs : ConfigurationSpace
         A ConfigurationSpace object that corresponds to the Tunable.
     """
+    # pylint: disable=too-complex
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/optimizers/convert_configspace.py:102]: still need this?

File: mlos_bench/mlos_bench/tunables/tunable.py [hunk 3]
```diff
@@ -182,19 +182,8 @@ def _sanity_check_numerical(self) -> None:
...
-                if self._quantization <= 1:
-                    raise ValueError(f"Number of quantization points is <= 1: {self}")
-            if self.dtype == float:
-                if not isinstance(self._quantization, (float, int)):
-                    raise ValueError(
-                        f"Quantization of a float param should be a float or int: {self}"
-                    )
-                if self._quantization <= 0:
-                    raise ValueError(f"Number of quantization points is <= 0: {self}")
+        if self._quantization is not None and self._quantization <= 1:
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tunables/tunable.py:185]:
```suggestion
        if self._quantization is not None and (not isinstance(self._quantization, int) or self._quantization <= 1):
```

File: mlos_bench/mlos_bench/tunables/tunable.py [hunk 4]
```diff
@@ -618,41 +606,45 @@ def quantized_values(self) -> Optional[Union[Iterable[int], Iterable[float]]]:
         """
         num_range = self.range
         if self.type == "float":
-            if not self._quantization:
+            if not self.cardinality:
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tunables/tunable.py:609]: This looks circular to me.  cardinality reads off of quantization and vice-versa?

Comment (bpkroth) [reply mlos_bench/mlos_bench/tunables/tunable.py:609]: Ah, nm, this is quantized values.  The diff got folded up in a confusing way.

File: mlos_bench/mlos_bench/tests/optimizers/toy_optimization_loop_test.py [hunk 5]
```diff
@@ -95,12 +95,12 @@ def test_mock_optimization_loop_no_defaults(
 def test_flaml_optimization_loop(mock_env_no_noise: MockEnv, flaml_opt: MlosCoreOptimizer) -> None:
     """Toy optimization loop with mock environment and FLAML optimizer."""
     (score, tunables) = _optimize(mock_env_no_noise, flaml_opt)
-    assert score == pytest.approx(60.15, 0.01)
+    assert score == pytest.approx(63.78, 0.01)
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/optimizers/toy_optimization_loop_test.py:98]: Is this a result of the change reported in #839 or the change in the quantization?

File: mlos_bench/mlos_bench/optimizers/convert_configspace.py (hunk 2 above)
Comment (motus) [reply mlos_bench/mlos_bench/optimizers/convert_configspace.py:102]: yes, unfortunately. we have one `if` too many in the method

File: mlos_bench/mlos_bench/tests/optimizers/toy_optimization_loop_test.py (hunk 5 above)
Comment (motus) [reply mlos_bench/mlos_bench/tests/optimizers/toy_optimization_loop_test.py:98]: no, this is a FLAML issue. I'll happily revert this change once we merge in the FLAML version fix

Comment (motus) [reply mlos_bench/mlos_bench/tests/optimizers/toy_optimization_loop_test.py]: done. I've rolled back that change. the unit test should pass once we merge in the FLAML version fix

File: mlos_bench/mlos_bench/tests/config/schemas/tunable-params/test-cases/bad/invalid/tunable-params-int-bad-float-quantization.jsonc [hunk 6]
```diff
@@ -6,7 +6,7 @@
                 "type": "float",
                 "default": 10,
                 "range": [1, 500],
-                "quantization": 0   // <-- should be greater than 0
+                "quantization": 0   // <-- should be greater than 1
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/config/schemas/tunable-params/test-cases/bad/invalid/tunable-params-int-bad-float-quantization.jsonc:9]:
```suggestion
                "quantization": 1   // <-- should be greater than 1
```

?

File: mlos_bench/mlos_bench/tests/tunables/tunable_to_configspace_test.py [hunk 7]
```diff
@@ -101,6 +103,9 @@ def configuration_space() -> ConfigurationSpace:
             TunableValueKind.RANGE,
         )
     )
+    hp = spaces["kernel_sched_latency_ns"]
+    assert isinstance(hp, NumericalHyperparameter)
+    _monkey_patch_quantization(hp, quantization_bins=10)
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/tunables/tunable_to_configspace_test.py:108]: Can we remove the `# TODO` above this?  Or is that waiting on #813 ?

File: mlos_bench/mlos_bench/tunables/tunable.py [hunk 8]
```diff
@@ -618,41 +606,45 @@ def quantized_values(self) -> Optional[Union[Iterable[int], Iterable[float]]]:
...
+            Either the number of points in the tunable or else None.
         """
         if self.is_categorical:
             return len(self.categories)
-        if not self.quantization and self.type == "float":
-            return np.inf
-        q_factor = self.quantization or 1
-        return int(self.span / q_factor) + 1
+        if self.quantization:
+            return self.quantization + 1
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tunables/tunable.py:644]: Is this accurate?

`quantization` is now equivalent to `number of bins` right?

Isn't that the same as `cardinality`?

File: mlos_bench/mlos_bench/optimizers/convert_configspace.py [hunk 9]
```diff
@@ -53,6 +49,34 @@ def _normalize_weights(weights: List[float]) -> List[float]:
...
+    Parameters
+    ----------
+    hp : NumericalHyperparameter
+        ConfigSpace hyperparameter to patch.
+    quantization_bins : int
+        Number of bins to quantize the hyperparameter into.
+    """
+    # Temporary workaround to dropped quantization support in ConfigSpace 1.0
+    # See Also: https://github.com/automl/ConfigSpace/issues/390
+    if not hasattr(hp, "sample_value_mlos_orig"):
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/optimizers/convert_configspace.py:65]:
```suggestion
    if quantization_bins <= 1:
        raise ValueError(f"Invalid quantization_bins: {quantization_bins}")

    if not hasattr(hp, "sample_value_mlos_orig"):
```

File: mlos_bench/mlos_bench/tests/tunables/test_tunables_size_props.py [hunk 10]
```diff
@@ -81,10 +86,17 @@ def test_tunable_quantized_float_size_props() -> None:
...
+            "range": [0, 1],
+            "default": 0,
+            "quantization": 10,
+        },
     )
     assert tunable.span == 1
-    assert tunable.cardinality == 11
+    # FIXME:
+    assert tunable.quantization == 10  # == tunable.quantization_num_bins == 11 ?
+    assert tunable.cardinality == 11  # == 10 ?
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/tunables/test_tunables_size_props.py:99]: @motus I put a few comments in here in my local version and pushed them here

File: mlos_bench/mlos_bench/tests/tunables/tunable_to_configspace_test.py (hunk 7 above)
Comment (motus) [reply mlos_bench/mlos_bench/tests/tunables/tunable_to_configspace_test.py:106]: yep, we can remove that comment now. the unit tests are here - although we probably should add a few more to #813 
I'll take a look once we merge in this one

File: mlos_bench/mlos_bench/tunables/tunable.py (hunk 8 above)
Comment (motus) [reply mlos_bench/mlos_bench/tunables/tunable.py:644]: I think you're right. I was trying to shoehorn the new quantization property semantics into the existing code and unit tests and did not realize how ugly that one-off quantization vs. cardinality now looks like. I'll fix it now.

File: mlos_bench/mlos_bench/tests/tunables/test_tunables_size_props.py (hunk 10 above)
Comment (motus) [reply mlos_bench/mlos_bench/tests/tunables/test_tunables_size_props.py:99]: yep, thank you! I'll also fix that quantization vs. cardinality issue now.

File: mlos_bench/mlos_bench/tunables/tunable.py (hunk 8 above)
Comment (motus) [reply mlos_bench/mlos_bench/tunables/tunable.py]: fixed!

**Review by bpkroth — APPROVED** [review]
LGTM, let's push the change in terminology (`quantized_num_bins` or whatever we decided) soon though.

**Pull request merged**